KIPs/kip-81.md

```diff
@@ -0,0 +1,773 @@
+4. After `trackEnd` block, the voting powers are frozen. The voting contract use StakingTracker getters to process votes casted by voter accounts.
+
+#### Example implementation
+
+Below is an example implementation of `createTracker()`, `refreshStake()`, and `refreshVoter()` functions.
+
+```solidity
+abstract contract StakingTracker {
+    function createTracker(uint256 trackStart, uint256 trackEnd)
+        external override returns(uint256 trackerId)
+    {
+        trackerId = allTrackerIds.length + 1;
+        allTrackerIds.push(trackerId);
+
+        Tracker storage tracker = trackers[trackerId];
+        tracker.trackStart = trackStart;
+        tracker.trackEnd = trackEnd;
+        populateFromAddressBook(tracker);
+        calcAllVotes(tracker);
+        return trackerId;
+    }
+
+    function populateFromAddressBook(Tracker storage tracker) private {
+        (address[] memory nodeIds,
+         address[] memory stakingContracts,
+         address[] memory rewardAddrs, , ) = IAddressBook(ADDRESS_BOOK_ADDRESS()).getAllAddressInfo();
+
+        // Group staking contracts by common reward address
+        for (uint256 i = 0; i < nodeIds.length; i++) {
+            address n = nodeIds[i];
+            address s = stakingContracts[i];
+            address r = rewardAddrs[i];
+            if (tracker.rewardToNodeId[r] == address(0)) { // fresh rewardAddr
+                tracker.nodeIds.push(n);
+                tracker.rewardToNodeId[r] = n;
+            } else { // previously appeared rewardAddr
+                n = tracker.rewardToNodeId[r];
+            }
+            tracker.stakingToNodeId[s] = n;
+            tracker.stakingBalances[s] = s.balance;
+            tracker.nodeBalances[n] += s.balance;
+        }
+    }
+
+    function calcAllVotes(Tracker storage tracker) private {
+        tracker.totalVotes = 0;
+        tracker.eligibleNodes = 0;
+        for (uint256 i = 0; i < tracker.nodeIds.length; i++) {
+            if (isNodeEligible(trackerId, tracker.nodeIds[i])) {
+                tracker.eligibleNodes ++;
+            }
+        }
+        for (uint256 i = 0; i < tracker.nodeIds.length; i++) {
+            address nodeId = tracker.nodeIds[i];
+            uint256 votes = calcVotes(tracker.eligibleNodes, tracker.nodeBalances[nodeId]);
+            tracker.nodeVotes[nodeId] = votes;
+            tracker.totalVotes += votes;
+        }
+    }
+
+    function calcVotes(uint256 eligibleNodes, uint256 balance) private view returns(uint256) {
+        uint256 voteCap = 1;
+        if (eligibleNodes > 1) {
+            voteCap = eligibleNodes - 1;
+        }
+        uint256 votes = balance / MIN_STAKE();
+        if (votes > voteCap) {
+            votes = voteCap;
+        }
+        return votes;
+    }
+
+    function refreshStake(address staking) external override {
+        for (uint256 i = 0; i < allTrackerIds.length; i++) {
+            uint256 trackerId = liveTrackerIds[i];
```

Comment:
liveTrackerIds is shown first here. `allTrackterIds` seems right?

Reply:
Fixed in https://github.com/klaytn/kips/commit/acfd52ebad8d79fd4e300bc034cf23dc501cba96.